### PR TITLE
P1679 string contains

### PR DESCRIPTION
**Pull request by WimLeflere**
C++23 string contains implementation

Paper: https://wg21.link/P1679R3
Draft issue: https://github.com/cplusplus/draft/issues/4328
Draft PR: https://github.com/cplusplus/draft/pull/4373

### CODE REVIEW

**Comment by mnatsuhara** [comment]
Thank you for your work in this PR!  However, I'd like to note that we won't be able to review this PR for a while, as the relevant issue #1450 says: 
> Note: We're still working on finishing C++20. Until we're done with that (and the compiler implements distinct /std:c++20 and /std:c++latest options), we won't be able to review PRs for C++23 features.

**Comment by WimLeflere** [comment]
My implementation is based on the `starts_with/ends_with` implementation.

I was wondering why the `const char*` function overload is noexcept.
`constexpr bool ends_with(const _Elem* const _Right) const noexcept /* strengthened */`

While this is explicitly not the case in the standard.

**Comment by StephanTLavavej** [comment]
The Standard historically avoided marking non-throwing but precondition-ful functions as `noexcept`. (The idea was that certain implementations might want to detect and report precondition violations by throwing an exception.) We often "strengthen" such functions to be `noexcept` in our implementation, with a comment (this is allowed by the Standard). Recently, the Library Working Group has changed its mind about this convention, although the entire Standard hasn't been updated accordingly yet.

**Comment by SuperWig** [comment]
Could be worth adding the explanation of the noexcept strengthening in the wiki.

**Comment by StephanTLavavej** [comment]
Now that #1929 has been merged, we can proceed with this PR, although `main` needs to be merged into it first.

**Comment by SuperWig** [comment]
Seems like @StephanTLavavej missed one

> This PR needs to merge main and resolve a conflict in yvals_core.h.

**Review by StephanTLavavej — COMMENTED** [review]
Thanks, this looks great! My apologies for the delay in reviewing your PR (due in part to C++20 marathon exhaustion and summer vacations). All I found were minor issues, so to save time I'll push changes (mainly adding `constexpr` for the `basic_string` overloads and updating a test accordingly) and move this to Final Review. After a second maintainer can double-check everything, we'll able to merge this for VS 2022 17.0 Preview 3.

**Review by StephanTLavavej — APPROVED** [review]

File: stl/inc/xstring [hunk 1]
```diff
@@ -4478,6 +4492,20 @@ public:
     }
 #endif // _HAS_CXX20
 
+#if _HAS_CXX23
+    _NODISCARD _CONSTEXPR20_CONTAINER bool contains(const basic_string_view<_Elem, _Traits> _Right) const noexcept {
+        return basic_string_view<_Elem, _Traits>(_Mypair._Myval2._Myptr(), _Mypair._Myval2._Mysize).contains(_Right);
```
Comment (miscco) [inline stl/inc/xstring:4497]: Nitpick, I believe we conventionally use `{}` to disambiguate function calls

Comment (StephanTLavavej) [reply stl/inc/xstring:4497]: Good catch - yep, that's the preferred convention in product code (although not all existing code follows it, especially for multiple-argument temporaries). @WimLeflere was following the precedent in `starts_with`/`ends_with` above which is definitely the right thing to do (local consistency is most important), but we can go ahead and change all of the occurrences in this file (more invasive cleanups should be kept separate from feature work but I think this is limited enough to be fine).

**Review by StephanTLavavej — APPROVED** [review]

**Review by CaseyCarter — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
Thanks for adding this feature to the Standard and implementing it here - and congratulations on your first microsoft/STL commit! :smile_cat: :tada: :rocket: This will ship in VS 2022 17.0 Preview 3.

**Pull request merged**